File: implementations/fcp-ref-lite/src/sil/integrity.ts
import { createHash } from 'node:crypto'
import { existsSync } from 'node:fs'
import { readFile, readdir } from 'node:fs/promises'
import { join } from 'node:path'
import type { Layout } from '../store/layout.js'
import { readJson, writeJson } from '../store/io.js'
import type { ImprintRecord } from '../boot/types.js'

export interface IntegrityDocument {
  version: '1.0'
  algorithm: 'sha256'
  files: Record<string, string>  // relative path → sha256 hash
}

export function sha256File(content: string): string {
  return 'sha256:' + createHash('sha256').update(content, 'utf8').digest('hex')
}

export function sha256Str(s: string): string {
  return 'sha256:' + createHash('sha256').update(s, 'utf8').digest('hex')
}

/**
 * Returns all files that must be tracked in the integrity document.
 * Covers: BOOT.md, baseline.json, skills/index.json, persona/*.md, skills/*\/manifest.json
 */
export async function trackedFiles(layout: Layout): Promise<string[]> {
  const paths: string[] = []

  if (existsSync(layout.bootMd)) paths.push(layout.bootMd)
  if (existsSync(layout.baseline)) paths.push(layout.baseline)
  if (existsSync(layout.skillsIndex)) paths.push(layout.skillsIndex)

  // persona/ files
  if (existsSync(layout.persona)) {
    const files = (await readdir(layout.persona))
      .filter(f => f.endsWith('.md'))
      .sort()
      .map(f => join(layout.persona, f))
    paths.push(...files)
  }

  // skills/*/manifest.json
  if (existsSync(layout.skills)) {
    const skillDirs = await readdir(layout.skills)
    for (const dir of skillDirs.sort()) {
      const manifestPath = join(layout.skills, dir, 'manifest.json')
      if (existsSync(manifestPath)) paths.push(manifestPath)
    }
  }

  return paths
}

/**
 * Compute SHA-256 hashes for all tracked files.
 * Returns a map of relative path → hash.
 */
export async function computeHashes(layout: Layout): Promise<Record<string, string>> {
  const files = await trackedFiles(layout)
  const result: Record<string, string> = {}
  for (const absPath of files) {
    const rel = absPath.startsWith(layout.root + '/')
      ? absPath.slice(layout.root.length + 1)
      : absPath
    const content = await readFile(absPath, 'utf8')
    result[rel] = sha256File(content)
  }
  return result
}

/**
 * Write state/integrity.json with current file hashes.
 */
export async function writeIntegrityDoc(layout: Layout): Promise<IntegrityDocument> {
  const files = await computeHashes(layout)
  const doc: IntegrityDocument = { version: '1.0', algorithm: 'sha256', files }
  await writeJson(layout.integrity, doc)
  return doc
}

/**
 * Verify current file hashes against integrity.json.
 * Returns list of drift descriptions (empty = clean).
 */
export async function verifyDrift(layout: Layout): Promise<string[]> {
  if (!existsSync(layout.integrity)) return ['integrity.json not found']

  const doc = await readJson<IntegrityDocument>(layout.integrity)
  const drifts: string[] = []

  for (const [rel, expected] of Object.entries(doc.files)) {
    const absPath = join(layout.root, rel)
    if (!existsSync(absPath)) {
      drifts.push(`missing: ${rel}`)
      continue
    }
    const content = await readFile(absPath, 'utf8')
    const actual = sha256File(content)
    if (actual !== expected) {
      drifts.push(`hash mismatch: ${rel}`)
    }
  }

  return drifts
}

export interface ChainVerificationResult {
  valid: boolean
  reason?: string
}

/**
 * Verify the integrity chain from GENESIS to last entry.
 *
 * Rules:
 * 1. First entry must be GENESIS with imprintHash matching genesisOmega in imprint.json
 * 2. Each subsequent entry's prevHash must equal sha256Str(JSON.stringify(previousEntry))
 * 3. Every ENDURE_COMMIT must have a non-empty evolutionAuthDigest
 *
 * An empty chain (entity has never evolved) is valid as long as imprint exists.
 */
export async function verifyChainFromImprint(layout: Layout): Promise<ChainVerificationResult> {
  if (!existsSync(layout.imprint)) {
    return { valid: false, reason: 'imprint.json not found' }
  }

  const imprint = await readJson<ImprintRecord>(layout.imprint)
  if (!imprint.genesisOmega) {
    return { valid: false, reason: 'imprint.json missing genesisOmega' }
  }

  // No chain yet — entity has not evolved, valid by definition
  if (!existsSync(layout.integrityChain)) {
    return { valid: true }
  }

  const { readJsonl } = await import('../store/io.js')
  const entries = await readJsonl<Record<string, unknown>>(layout.integrityChain)

  if (entries.length === 0) return { valid: true }

  // Rule 1: first entry must be GENESIS with correct imprintHash
  const first = entries[0]!
  if (first['type'] !== 'GENESIS') {
    return { valid: false, reason: `first chain entry is not GENESIS (got ${first['type']})` }
  }
  if (first['data'] == null || (first['data'] as Record<string, unknown>)['imprintHash'] !== imprint.genesisOmega) {
    return { valid: false, reason: 'GENESIS imprintHash does not match genesisOmega' }
  }

  // Rules 2 & 3: traverse remaining entries
  let prevEntry = first
  for (let i = 1; i < entries.length; i++) {
    const entry = entries[i]!
    const expectedPrevHash = sha256Str(JSON.stringify(prevEntry))

    // Rule 2: prevHash linkage
    if (entry['prevHash'] !== expectedPrevHash) {
      return {
        valid: false,
        reason: `chain broken at seq ${entry['seq']}: prevHash mismatch`,
      }
    }

    // Rule 3: ENDURE_COMMIT must have evolutionAuthDigest
    if (entry['type'] === 'ENDURE_COMMIT') {
      const data = entry['data'] as Record<string, unknown> | undefined
      if (!data?.['evolutionAuthDigest']) {
        return {
          valid: false,
          reason: `ENDURE_COMMIT at seq ${entry['seq']} missing evolutionAuthDigest`,
        }
      }
    }

    prevEntry = entry
  }

  return { valid: true }
}

File: experiments/fcp-ref-python/fcp_base/sil/integrity.py
"""
SIL structural verification — §10.1.

Tracks structural files, computes and verifies SHA-256 integrity documents,
and validates the integrity chain (including evolution auth coverage).
"""

from __future__ import annotations

import json
from typing import TYPE_CHECKING

from ..acp import parse_envelope_data
from ..formats import ChainEntry, ChainEntryType, IntegrityDocument
from ..store import atomic_write, read_jsonl
from .utils import sha256_file, sha256_str

if TYPE_CHECKING:
    from ..store import Layout


def tracked_files(layout: "Layout") -> list:
    """Return all files that must be present in the Integrity Document."""
    from pathlib import Path
    paths: list[Path] = [layout.boot_md, layout.baseline, layout.skills_index]

    if layout.persona_dir.is_dir():
        paths.extend(sorted(layout.persona_dir.iterdir()))

    for manifest in sorted(layout.skills_dir.glob("*/manifest.json")):
        paths.append(manifest)
    for manifest in sorted(layout.skills_lib_dir.glob("*/manifest.json")):
        paths.append(manifest)

    return [p for p in paths if p.is_file()]


def compute_integrity_files(layout: "Layout") -> dict[str, str]:
    """Compute SHA-256 hashes for all tracked structural files."""
    result: dict[str, str] = {}
    for p in tracked_files(layout):
        rel = str(p.relative_to(layout.root))
        result[rel] = sha256_file(p)
    return result


def write_integrity_doc(layout: "Layout", files: dict[str, str]) -> None:
    """Write state/integrity.json atomically with the given file hashes."""
    doc = IntegrityDocument(
        version="1.0",
        algorithm="sha256",
        last_checkpoint=None,
        files=files,
    )
    atomic_write(layout.integrity_doc, doc.to_dict())


def verify_structural_files(
    layout: "Layout", integrity_doc: IntegrityDocument
) -> list[str]:
    """Recompute hashes and compare against *integrity_doc*.

    Returns a list of mismatch descriptions.  Empty list means clean.
    """
    mismatches: list[str] = []
    for rel, expected in integrity_doc.files.items():
        p = layout.root / rel
        if not p.exists():
            mismatches.append(f"missing: {rel}")
            continue
        actual = sha256_file(p)
        if actual != expected:
            mismatches.append(f"hash mismatch: {rel}")
    return mismatches


def verify_integrity_chain(layout: "Layout", integrity_doc: IntegrityDocument) -> bool:
    """Validate the integrity chain from the last checkpoint forward.

    Returns True if valid, False if any gap, hash mismatch, or missing
    authorization reference is detected.
    """
    if not layout.integrity_chain.exists():
        return False

    entries = read_jsonl(layout.integrity_chain)
    if not entries:
        return False

    cp = integrity_doc.last_checkpoint
    start_seq = cp.seq if cp is not None else 0

    lines = layout.integrity_chain.read_text(encoding="utf-8").splitlines()
    seq_to_line: dict[int, str] = {}
    for line in lines:
        line = line.strip()
        if not line:
            continue
        d = json.loads(line)
        seq_to_line[d["seq"]] = line

    check_from = 1 if cp is None else start_seq
    prev_hash: str | None = None

    for seq in sorted(seq_to_line.keys()):
        if seq < check_from:
            continue
        raw = seq_to_line[seq]
        entry = ChainEntry.from_dict(json.loads(raw))

        if entry.type == ChainEntryType.GENESIS:
            if entry.imprint_hash is None:
                return False
        else:
            if prev_hash is not None and entry.prev_hash != prev_hash:
                return False
            if entry.type == ChainEntryType.ENDURE_COMMIT:
                if not entry.evolution_auth_digest:
                    return False

        prev_hash = sha256_str(raw)

    cp_final = cp
    if cp_final is not None:
        raw_cp = seq_to_line.get(cp_final.seq)
        if raw_cp is None:
            return False
        actual_digest = sha256_str(raw_cp)
        if actual_digest != cp_final.digest:
            return False

    if not _verify_evolution_auth_coverage(layout, entries):
        return False

    return True


def _verify_evolution_auth_coverage(layout: "Layout", chain_entries: list[dict]) -> bool:
    """Check that each ENDURE_COMMIT in the chain has a matching EVOLUTION_AUTH."""
    auth_digests: set[str] = set()
    if layout.integrity_log.exists():
        for log_entry in read_jsonl(layout.integrity_log):
            data = parse_envelope_data(log_entry)
            if data.get("type") != "EVOLUTION_AUTH":
                continue
            digest = data.get("auth_digest", "")
            if digest:
                auth_digests.add(digest)

    for raw_entry in chain_entries:
        entry = ChainEntry.from_dict(raw_entry)
        if entry.type != ChainEntryType.ENDURE_COMMIT:
            continue
        if not entry.evolution_auth_digest:
            return False
        if entry.evolution_auth_digest not in auth_digests:
            return False

    return True
